### PR TITLE
Ruby: type-tracking and API edges through simple library callables

### PR DESCRIPTION
**Pull request by asgerf**
This PR ensures that simple library callables whose callee does not depend on API graphs can be seen by type-tracking and API graphs. "Simple" in this case means it can be reduced to a store or load edge at the call site, though in the future we may want to add load-store and basic edges as well.

For example, a summary of form `Argument[0].Element[0];ReturnValue` will be seen as:
- a load step in type-tracking, reading the content set corresponding to `Element[0]`, and
- an edge in the API graph labelled with the content corresponding to `Element[0]`.

Ruby is the first language to use the string-based flow summary approach combined with type-tracking and API graphs, but the two systems hadn't been fully integrated. In particular, Ruby has so far been modelling the standard library using summaries that are invisible to type-tracking and API graphs - this would lead to incompleteness compared to just generating the flow edges directly. I believe this PR bridges the gap, so we can get the best of both worlds while modelling everything with flow summaries.

[Evaluation](https://github.com/github/codeql-dca-main/issues/7203) shows 216 new call edges, at a modest performance cost. I've investigated the performance overhead a bit, but couldn't get it any lower than it is now - partly because we are just tracking more flow now.

I'm running an evaluation for Python as well, since some shared code was affected.

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTracker.qll
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/typetracking/TypeTracker.qll:79]: This should be `getAReadContent`.

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll:153]: You should be able to use `summarySetterStep` from `FlowSummaryImpl` instead.

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll:193]: You should be able to use `summaryGetterStep` from `FlowSummaryImpl` instead.

File: ruby/ql/test/library-tests/modules/calls.rb
Comment (hvitved) [inline ruby/ql/test/library-tests/modules/calls.rb:129]: Great that this is fixed, why is that?

**Review by hvitved — CHANGES_REQUESTED** [review]
Looks great Asger! A few comments.

File: ruby/ql/lib/codeql/ruby/ApiGraphs.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/ApiGraphs.qll:265]: It is done like this exactly to avoid the fan-out from `getAReadContent`.

Here's a break down of how the graph is generated and what ends up being matched:
| Description | Edge label  | Matched by  |
|---|---|---|
| Use node reading a known element `n`   | `TKnownElementContent(n)`  | `Element[n]` and `Element[any]`  |
| Def node storing a known element `n`   | `TKnownElementContent(n)`  | `Element[n]` and `Element[any]`  |
| Use node reading an unknown element   | `TUnknownElementContent`  | `Element[?]` and `Element[any]`  |
| Def node storing an unknown element   | `TUnknownElementContent`  | `Element[?]` and `Element[any]`  |

`Element[n]` does not match `TUnknownElementContent`, and the model must thus use `Element[?,n]` if this behaviour is wanted. This seems consistent with how existing Ruby models are required to mention `Element[?]` explicitly.

I believe this is the best integration we can hope for without changing how `ContentSet` works.

I tried storing `ContentSet` on the edge label but it doesn't really work out nicely. `getAReadContent()` returns a `Content`, but the edges are labelled with a `ContentSet` so we have to map it back to a `ContentSet` to look up in the `edge` relation. To ensure an efficient join there we'd end up factoring out a relation indexed on the `Content`, which basically corresponds the edge relation we're using now.

File: ruby/ql/lib/codeql/ruby/ApiGraphs.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/ApiGraphs.qll:542]: After applying the changes in this PR, it didn't perform well anymore.

The pipeline looked like this on `openproject` (manually aborted after a while)
```
(310s) Tuple counts for ApiGraphs#3116a2f3::API::Impl::useNodeStep#3#ffff/4@022f8a1e after 3m48s:
189468     ~0%      {2} r1 = JOIN ApiGraphs#3116a2f3::API::Impl::useCandRev#1#ff_10#join_rhs WITH DataFlowPrivate#462ff392::Cached::isLocalSourceNode#1#f ON FIRST 1 OUTPUT Lhs.1, Lhs.0 'result'
189468     ~0%      {3} r2 = JOIN r1 WITH num#TypeTracker#6f8aa30b::MkTypeBackTracker#fff_2#join_rhs ON FIRST 1 OUTPUT Lhs.1 'result', Lhs.1 'result', Lhs.0
51903      ~3%      {4} r3 = JOIN r2 WITH TypeTracker#6f8aa30b::Cached::stepCall#3#fff_102#join_rhs ON FIRST 1 OUTPUT Lhs.2, Lhs.1 'result', Rhs.1 'mid', Rhs.2
49779      ~2%      {4} r4 = JOIN r2 WITH TypeTracker#6f8aa30b::Cached::stepNoCall#3#fff_102#join_rhs ON FIRST 1 OUTPUT Lhs.2, Lhs.1 'result', Rhs.1 'mid', Rhs.2
101682     ~0%      {4} r5 = r3 UNION r4
189161     ~6%      {4} r6 = JOIN r5 WITH TypeTracker#6f8aa30b::TypeBackTracker::getACompatibleTypeTracker#0#dispred#ff ON FIRST 1 OUTPUT Lhs.3, Rhs.1 't', Lhs.1 'result', Lhs.2 'mid'
1322764717 ~12%     {4} r7 = JOIN r6 WITH TypeTracker#6f8aa30b::Cached::append#2#fff_120#join_rhs ON FIRST 2 OUTPUT Lhs.3 'mid', Rhs.2 'tmid', Lhs.1 't', Lhs.2 'result'
                    return r7
```

When I reverted to idiomatic type tracking, it seemed to perform quite well again, as witnessed by the performance evaluation.

[This evaluation](https://github.com/github/codeql-dca-main/tree/data/asgerf/summarize-loads-trackUseNodeComparison/reports) shows the effect of the commit that reverts to idiomatic type tracking (i.e. the whole branch without that commit, versus the brach with that commit).

We could of course try to hack on the join ordering again, but I couldn't find an example where the pruning seemed beneficial. If you remember a specific project where the pruning was important, please let me know and I'll investigate.

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTracker.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/typetracking/TypeTracker.qll:74]: As discussed offline, this was done to avoid the fan-out in `getAReadContent()`. However, it caused false negatives when backtracking. I have added a test that produces the missed backtracking flow.

I have changed type trackers to store a `ContentSet` instead, and added `compatibleContents(ContentSet, ContentSet)` to explicitly match up the content sets of a store and a load.

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTracker.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/typetracking/TypeTracker.qll:79]: See answer above.

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll:153]: That gives us a step inside the synthesised callable, whereas we need a step at the call site.

If we start doing bottom-up summarisation of type-tracking steps then we could use `summarySetterStep` instead. However, API graphs would still only want the step at the original call site so the logic is still needed somewhere.

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll:193]: See answer above.

File: ruby/ql/test/library-tests/modules/calls.rb
Comment (asgerf) [reply ruby/ql/test/library-tests/modules/calls.rb:129]: Turns out it was fixed for the wrong reasons. The summary seems to depend on having store-load steps so I'll revisit it when adding those (which are relatively simple in type-tracking but more complex in API graphs, so I'd rather not handle it in this PR).

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll:395]: Is it a problem to have it here? I don't really see how it would help to have a new union type for this.

Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll]: Actually now that type-trackers contain `ContentSet` instead of `Content`, it's actually not that easy to factor out anymore.

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll]: Moved to `TypeTrackerSpecific` 👍

**Comment by asgerf** [comment]
[New evaluation](https://github.com/github/codeql-dca-main/issues/7203)

File: ruby/ql/lib/codeql/ruby/ApiGraphs.qll
Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/ApiGraphs.qll:542]: FWIW, `useNodeStep` only [got introduced recently](https://github.com/github/codeql/pull/10141), and I was never convinced that it was right to join with `append` first. I wonder what it would look like if that PR got reverted.

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll
Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll:162]: `summarySetterStep` is a step at the call site.

File: ruby/ql/lib/codeql/ruby/ApiGraphs.qll
Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/ApiGraphs.qll:265]: Thanks for the detailed explanation.

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll:386]: Couldn't this rather be defined as a `newtype` inside `TypeTracker.qll`:
```ql
newtype TOptionalContentSet =
  TContentSetSome(ContentSet cs) or
  TContentSetNone()
```

File: ruby/ql/lib/codeql/ruby/ApiGraphs.qll
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/ApiGraphs.qll:271]: I wouldn't expect `getAReadContent` to have large fan-in (only large fan-out).

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll:386]: You're asking if I want to revert https://github.com/github/codeql/pull/10375/commits/b1c18986483520926ea004c81d7b8b7bebbef983?

Repeatedly boxing values is just less efficient than putting them in the same universe. The `Optional` wrapper is the most obvious one to avoid.

File: ruby/ql/lib/codeql/ruby/ApiGraphs.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/ApiGraphs.qll:271]: As discussed in chat, this just gives a better join order, it's not that it's too large to materialize.

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll:318]: I'm not sure which class you're referring to
- `TOptionalContentSet` is entirely hidden in `DataFlowPrivate`.
- `OptionalTypeTrackerContent` is defined in `TypeTrackerSpecific.qll`
- `Content::Attribute` is not hidden as it can be wrapped in a `ContentSet`. We could hide the class, but the values can still leak out to the user.

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll:162]: Ah, sorry I was too a bit quick to conclude what was going on in that predicate.

It does seem to include taint steps though, which we don't want here.

Would you prefer if I add `boolean valuePreserving` as a parameter to that predicate so we can use it from here?

File: ruby/ql/lib/codeql/ruby/ApiGraphs.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/ApiGraphs.qll:542]: Reverting my commit and then reverting that PR seems to handle `openproject` without blowing up, albeit with larger tuple counts than with idiomatic type-tracking. I've [started an evaluation](https://github.com/github/codeql-dca-main/issues/7299) of the two branches to see which performs better.

**Comment by asgerf** [comment]
We've got quite a few unresolved threads open. @hvitved could you 'mark as resolved' any conversations you feel have been addressed.

From my perspective the two outstanding items are:
- [trackUseNode pruning](https://github.com/github/codeql/pull/10375#discussion_r979958693): Waiting for [this evaluation](https://github.com/github/codeql-dca-main/issues/7299).
- [summarySetterStep](https://github.com/github/codeql/pull/10375#discussion_r979890000): Waiting for your feedback in  on whether to add `boolean valuePreserving` to the relevant predicates.

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll
Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll:173]: 👍

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll
Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll:213]: Remember to also update this bit.

**Comment by hvitved** [comment]
> @hvitved could you 'mark as resolved' any conversations you feel have been addressed.

Done.

File: ruby/ql/lib/codeql/ruby/ApiGraphs.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/ApiGraphs.qll]: Evaluation was messed up, trying again [here](https://github.com/github/codeql-dca-main/issues/7328).

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll:173]: Unfortunately this leads to non-monotonic recursion, since that predicate ultimately depends on `SummarizedCallable.getACall` whereas we need something that only depends on `getACallSimple`.

File: ruby/ql/lib/codeql/ruby/ApiGraphs.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/ApiGraphs.qll]: [Evaluation](https://github.com/github/codeql-dca-main/issues/7334) comparing pruning with https://github.com/github/codeql/pull/10141 reverted, versus idiomatic type-tracking with no pruning.

The evaluation does not show any benefit from pruning. It's possible that it originally capitalised on incompleteness, and as the call graph and API graph are now more complete, many more nodes can now be tracked to a use site.

**Comment by asgerf** [comment]
Force-pushed to remove some revert commits that I had accidentally pushed as part of the investigation in [this thread](https://github.com/github/codeql/pull/10375#discussion_r979958693).

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll
Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll:173]: Ah, right.

**Comment by asgerf** [comment]
Rebased to fix conflict in `DataFlowDispatch.qll`: the call to `getACallSimple` now appears in `viableLibraryCallable`.

File: ruby/ql/lib/codeql/ruby/ApiGraphs.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/ApiGraphs.qll:265]: After merging with https://github.com/github/codeql/pull/10574 I believe the table is updated as follows:
- the first two rows can also be matched by `Element[n!]`
- the last two rows can also be matched by `Element[k]` for any constant `k`

| Description | Edge label  | Matched by  |
|---|---|---|
| Use node reading a known element `n`   | `TKnownElementContent(n)`  | `Element[n]`, `Element[n!]`, and `Element[any]`  |
| Def node storing a known element `n`   | `TKnownElementContent(n)`  | `Element[n]`, `Element[n!]`, and `Element[any]`  |
| Use node reading an unknown element   | `TUnknownElementContent`  | `Element[k]` for any `k`, `Element[?]` and `Element[any]`  |
| Def node storing an unknown element   | `TUnknownElementContent`  | `Element[k]` for any `k`, `Element[?]` and `Element[any]`  |

- Previously it was the model's responsibility to mention `Element[?]` if unknown reads/stores should be found.
- Now models are now responsible for using `Element[n!]` if this behaviour is _not_ wanted.

Again this seems consistent with how Ruby models work after https://github.com/github/codeql/pull/10574, so I'm cautiously optimistic that everything just works after this change.

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll:388]: Perhaps add a `// Only used by type-tracking` comment here as well.

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll:274]: I think this can be replaced with
```ql
argumentPositionMatch(call, result, pos)
```

**Review by hvitved — COMMENTED** [review]
Two last comments (sorry for not catching earlier).

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll]: Thanks! 👍  I was a little sad about having to do this here.

**Comment by asgerf** [comment]
The evaluation shows a slightly increased cost for ruby/ruby after merging with https://github.com/github/codeql/pull/10574. I think part of the reason is that there are about twice as many content sets now, and ruby/ruby has a lot of content sets. I tried restricting the content sets we care about with [this commit](https://github.com/asgerf/codeql/commit/1c4c05810fc9a3d1f893b3b07d9781bfbf13c55e) and although the tuple counts showed a great reduction, [it did not seem to help](https://github.com/github/codeql-dca-main/tree/data/asgerf/rb/summarize-loads-v2-tt-domains/reports). I'd prefer to move ahead with the PR as it is.

**Comment by hvitved** [comment]
> I tried restricting the content sets we care about with [this commit](https://github.com/asgerf/codeql/commit/1c4c05810fc9a3d1f893b3b07d9781bfbf13c55e) and although the tuple counts showed a great reduction, [it did not seem to help](https://github.com/github/codeql-dca-main/tree/data/asgerf/rb/summarize-loads-v2-tt-domains/reports). I'd prefer to move ahead with the PR as it is.

I do like that commit, in particular because `TStepSummary` is cached. We could even make the two branches even smaller via
```ql
private predicate compatibleStoreLoadContents(TypeTrackerContent storeContents, TypeTrackerContent loadContents) {
  basicStoreStep(_, _, storeContents) and
  basicLoadStep(_, _, loadContents) and
  compatibleContents(storeContents, loadContents)
}
...
    StoreStep(TypeTrackerContent content) { compatibleStoreLoadContents(content, _)  } or
    LoadStep(TypeTrackerContent content) { compatibleStoreLoadContents(_, content) } or
```

**Comment by asgerf** [comment]
I measured the size of `StepSummary` with each change on `ruby/ruby`:

| Commit | Size |
|--------|--------|
| Baseline dc03557 | 4,111,472 |
| Size restriction from https://github.com/asgerf/codeql/commit/1c4c05810fc9a3d1f893b3b07d9781bfbf13c55e | 4,633 |
| Additional size restriction from @hvitved's comment | 4,428 |

Unfortunately neither version is compatible with `TypeTracker.startInContent` as a model might start type-tracking in a property that has a load in the database but not a store. The predicate isn't currently used in Ruby, but has uses in JS and Python.

I've pushed another commit that makes the restriction on type-tracker compatible with `startInContent` and start another evaluation (not that I expect it look better than last time).

**Review by hvitved — APPROVED** [review]
Thanks for you hard work, and patience, on this one!

**Comment by asgerf** [comment]
[Evaluation](https://github.com/github/codeql-dca-main/tree/data/asgerf/summarize-loads__nightly__code-scanning__25/reports) looks similar to the last one.

The [last Python evaluation](https://github.com/github/codeql-dca-main/tree/data/asgerf/summarize-loads-python/reports) shows a median slow-down of 1% which is a bummer given that this PR does nothing for Python. We can't do langauge-dependent join order tweaks in shared code, and the best join order is slightly different since Ruby has `ContentSet` and Python doesn't.

**Comment by asgerf** [comment]
[Another evaluation](https://github.com/github/codeql-dca-main/issues/7444) of Python including the latest changes shows the same result, and I got a 👍 from the Python team that we're ok with this.

**Pull request merged**
